### ISSUE TITLE
Jakarta EE 10: Support Jakarta Contexts and Dependency Injection 4.0

### ISSUE THREAD

**Issue by jhanders34** [OpenLiberty/open-liberty#18728]
## Description of the high level feature, including any external spec links:  
Implement the Jakarta Contexts and Dependency Injection 4.0 specification

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## When complete & mandatory, add links to the UFO (Upcoming Feature Overview) document, FTS (Feature Test Summary), and blogs post issues(s):
- UFO: https://ibm.box.com/s/ub983af3yzskf3ev6g27ehb3e2avb08k
- FTS: https://github.com/OpenLiberty/open-liberty/issues/19870
- Beta Blog Post (if applicable): https://github.com/OpenLiberty/open-liberty/issues/21777
- Blog Post: https://github.com/OpenLiberty/open-liberty/issues/24501

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## List of Steps to complete or get approvals / sign-offs for Onboarding to the Liberty release (GM date)

Instructions:
- Do the actions below and mark them complete in the checklist when they are done.
- Make sure all feature readiness approvers put the appropriate tag on the epic to indicate their approval.

## **Design**
Before Development Starts or 8 weeks before Onboarding
- [ ] POC Design / UFO Review Scheduled (David Chang) or N/A.
- [ ] POC Design / UFO Reviewed (Feature Owner) or N/A.
- [ ] Complete any follow-ons from the POC Review.
- [ ] Design / UFO Approval ([Alasdair Nottingham](https://github.com/NottyCode)) or N/A.
- [ ] No Design / No UFO Approval ([Alasdair Nottingham](https://github.com/NottyCode)) or N/A.
- [ ] SVT Requirements identified. (Epic owner / Feature owner with SVT focal point)
- [ ] ID Requirements identified ([Documenting Open Liberty](https://github.com/OpenLiberty/open-liberty/wiki/Documenting-Open-Liberty)). (Epic owner / Feature owner with ID focal point)
- [ ] Create a child task of this epic entitled "Feature Test Summary" via [this template](https://github.com/OpenLiberty/open-liberty/issues/new?assignees=&labels=Feature+Test+Summary&template=feature_test_summary.md&title=). Add the link in above.

##
##

## **Development**
When active development has begun 
- [x] Add the "In Progress" label to this issue.
- [x] #19910 
- [x] #19748 
- [x] #19912
- [x] #19868 
- [x] #20173
- [x] #19867
- [x] #19900 
- [x] #20570
- [x] #19870 
- [x] #19869 
- [x] #19911 

## **Beta**
If your feature, or portions of it, are going to be included in a beta  
Before Onboarding the beta
- [x] Beta Fence the functionality (`kind=beta`, `ibm:beta`, `ProductInfo.getBetaEdition()`)  

1 week before beta GA
- [x] Create, populate, and link to the [Beta blog post issue](https://github.com/OpenLiberty/open-liberty/issues/new?assignees=lauracowen%2C+jakub-pomykala&labels=&template=blog_post_beta.md&title=BETA+BLOG+-+title_of_your_update)

## **Legal**
3 weeks before Onboarding
- [ ] Identify all open source libraries that are changing or are new. Work with Legal Release Services (Cass Tucker or Release PM) to get open source cleared and approved. Or N/A. (Epic Owner).   New or changed open source impacts license and Certificate of Originality.

## **Translation**
3 weeks before Onboarding
- [x] All new or changed PII messages are checked into the integration branch, before the last translation shipment out. (Epic Owner)

## **Feature Complete**
2 weeks before Onboarding
- [x] Implementation complete. (Epic owner / Feature owner)
- [ ] All function tests complete. Ready for FAT Approval. (Epic owner / Feature owner)
- [ ] Review all known issues for Stop Ship. (Epic owner / Feature owner / PM)

## **Focal Point Approvals**
2 to 1 week before Onboarding
#### You **MUST** have the Design Approved or No Design Approved label before requesting focal point approvals.

All features (both "Design Approved" and "No Design Approved")
- [x] **FAT** - ([OpenLiberty/fat-approvers](https://github.com/orgs/OpenLiberty/teams/fat-approvers)). SOE FATS are running successfully or N/A . Approver adds label focalApproved:fat to the Epic in Github.
- [x] **Demo** - ([Tom Evans](https://github.com/tevans78) or [Chuck Bridgham](https://github.com/cbridgha)). Demo is scheduled for an upcoming EOI. Approver adds label focalApproved:demo to the Epic in Github.
- [x] **Globalization** ([Sam Wong](https://github.com/samwatibm) - Liberty / Simy Cheeran - tWAS). Translation is complete or N/A. TVT - complete or N/A. Approver adds label focalApproved:globalization to the Epic in Github.

"Design Approved" features
- [x] **Accessibility** - ([Steven Zvonek](https://github.com/steven1046)). Accessibility testing is complete or N/A. Approver adds label focalApproved:accessibility to the Epic in Github.
- [x] **ID** - ([Karen Deen](https://github.com/chirp1)). Documentation work is complete or N/A . Approver adds label focalApproved:id to the Epic in Github.
- [x] **Performance** - ([Jared Anderson](https://github.com/jhanders34)). Performance testing is complete with no high severity defects or N/A . Approver adds label focalApproved:performance to the Epic in Github.
- [x] **Serviceability** - ([Don Bourne](https://github.com/donbourne)). Serviceability has been addressed.
- [x] **STE** - (Swati Kasundra). STE chart deck is complete or N/A . Approver adds label focalApproved:ste to the Epic in Github.
- [x] **SVT** - ([Brian Hanczaryk](https://github.com/hanczaryk) - APS). SVT is complete or N/A . Approver adds label focalApproved:svt to the Epic in Github.

## **Ready for GA**
1 week before Onboarding
- [ ] No Stop Ship issues for the feature. (Epic owner / Feature owner / Release PM)
- [ ] Ship Readiness Review and Release Notes completed (Epic owner / Feature owner / Release PM)
- [ ] Github Epic and Epic's issues are closed / complete. All PRs are committed to the release branch. (Epic owner / Feature owner / Backlog Subtribe PM)

1 week before GA
- [ ] Create, populate, and link to the [Blog post issue](https://github.com/OpenLiberty/open-liberty/issues/new?assignees=lauracowen%2C+jakub-pomykala&labels=&template=blog_post_ga_release.md&title=GA+BLOG+-+title_of_your_update)

## **Other deliverbles**
- [ ] **OL Guides** - ([Yee-Kang Chang](https://github.com/yeekangc)). Assessment for OL Guides is complete or N/A.
- [ ] **WDT** - (Leonard Theivendra). WDT work complete or N/A.
- [ ] **Blog** - ([Laura Cowen](https://github.com/lauracowen)) Blog article writeup (Epic owner / Feature owner / Laura Cowen)
- [x] **OL Docs** - https://github.com/OpenLiberty/docs/issues/5754

**Comment by kwsutter** [comment]
Just linking this Epic with the CDI Lite Epic #18729.

**Comment by tevans78** [comment]
Will aim to do this issue as part of CDI 4.0 - https://github.com/OpenLiberty/open-liberty/issues/18188

**Comment by garypicher** [comment]
UFO Review Part 1 follow-ups:

1. Make sure there's a test to verify that CDI runtime is initialized and the Startup notifications sent before any business logic is able to executive.  (The specific concern is with startup beans.)
- TE: Added a test here https://github.com/OpenLiberty/open-liberty/pull/21057. However it should be noted that `@PostConstruct` methods of a `@Startup` EJB are fired before the "Startup" CDI events are triggered. We do not believe that this is a problem.

2. The "emptyBeansXMLExplicitBeanArchive" attribute may trigger a test failure that checks for attribute names that are too long.  This can be overridden but should still check with Alasdair to ensure that we are okay with this name.
- TE: Issue opened https://github.com/OpenLiberty/open-liberty/issues/21019

3. Need to schedule a UFO Review Part 2 to finish the review.

**Comment by yeekangc** [comment]
From UFO Review Part 2:
- Slide 42: Normal precedence is to not warn if the config attributes don't apply to the version in use.
-- TE: https://github.com/OpenLiberty/open-liberty/issues/21094

- Slide 43: Change CWOWB1015 into an INFO message.  Shouldn't warn (out of zero migration consideration).
-- TE: https://github.com/OpenLiberty/open-liberty/issues/21095

- Slide 44/45: Use of cdi vs cdi12 presents a special case for our Dev Tools to handle particularly the Language Server for Liberty Config.
- Slide 55: Consider what we can potentially do in Dev Tools to help developers with the change in meaning of empty beans.xml.  And/or, consider how we can better handle this case to prevent users from hitting issues by accident (e.g., by checking for it).
- Slide 56: Potential for LSP4Jakarta to check for `@Nonnull` and `@Nullable` though it may not be perfect.

-- TE: The UFO has been updated to mention the potential updates to our Dev Tools and LSP4Jakarta. @yeekangc will open suitable issues and/or epics.

**Comment by tevans78** [comment]
@NottyCode Now that design issue https://github.com/OpenLiberty/open-liberty/issues/21019 has reached a conclusion, I have updated the UFO and I think it is ready for approval.

**Comment by donbourne** [comment]
**Serviceability Approval Comment** - Please answer the following questions for serviceability approval:

1.  UFO -- does the UFO identify the most likely problems customers will see and identify how the feature will enable them to diagnose and solve those problems without resorting to raising a PMR?  Have these issues been addressed in the implementation?
> Yes, the most likely issues are around the change in meaning of an empty beans.xml

2.  Test and Demo -- As part of the serviceability process we're asking feature teams to test and analyze common problem paths for serviceability and demo those problem paths to someone not involved in the development of the feature (eg. L2, test team, or another development team).  
a) What problem paths were tested and demonstrated?
> We looked at the various new problem paths around beans.xml

b) Who did you demo to?
> Ben Confino

c) Do the people you demo'd to agree that the serviceability of the demonstrated problem scenarios is sufficient to avoid PMRs for any problems customers are likely to encounter, or that L2 should be able to quickly address those problems without need to engage L3?
> Yes

3.  SVT -- SVT team is often the first team to try new features and often encounters problems setting up and using them.  Note that we're not expecting SVT to do full serviceability testing -- just to sign-off on the serviceability of the problem paths they encountered.
a) Who conducted SVT tests for this feature?
> Brian Hanczaryk

b) Do they agree that the serviceability of the problems they encountered is sufficient to avoid PMRs, or that L2 should be able to quickly address those problems without need to engage L3?
> Jakarta EE 10 testing is still ongoing 

4.  Which L2 / L3 queues will handle PMRs for this feature?  Ensure they are present in the contact reference file and in the queue contact summary, and that the respective L2/L3 teams know they are supporting it.  Ask Don Bourne if you need links or more info.
> WAS L3: CDI

5.  Does this feature add any new metrics or emit any new JSON events?  If yes, have you updated the JMX metrics reference list / Metrics reference list / JSON log events reference list in the Open Liberty docs?
> Nope

**Comment by cbridgha** [comment]
Please add a demo to one of the upcoming EOI meetings to get demo approval - @tevans78

**Comment by tevans78** [comment]
@cbridgha EOI Demo scheduled for WebSphere EOI 22.22 - 15th November 2022

**Comment by donbourne** [comment]
marking this as approved for serviceability as I don't expect any issues due to nature of changes in CDI 4.0, and lack of any serviceability comments/issues raised on this so far.

**Comment by chirp1** [comment]
Info to document received at https://github.com/OpenLiberty/docs/issues/5754. Approving the epic.

**Comment by tevans78** [comment]
All done, closing the Epic ;-)
